target/linux/generic/pending-5.10/499-mtd-spi-nor-variable-erase.patch

```diff
@@ -0,0 +1,204 @@
+ drivers/mtd/spi-nor/Kconfig | 10 +++++++
+ drivers/mtd/spi-nor/core.c  | 10 +++++--
+ include/linux/mtd/mtd.h     |  2 ++
+ 4 files changed, 60 insertions(+), 14 deletions(-)
+
+--- a/drivers/mtd/mtdpart.c
++++ b/drivers/mtd/mtdpart.c
+@@ -40,10 +40,11 @@ static struct mtd_info *allocate_partiti
+ 	struct mtd_info *master = mtd_get_master(parent);
+ 	int wr_alignment = (parent->flags & MTD_NO_ERASE) ?
+ 			   master->writesize : master->erasesize;
++	int wr_alignment_minor = 0;
+ 	u64 parent_size = mtd_is_partition(parent) ?
+ 			  parent->part.size : parent->size;
+ 	struct mtd_info *child;
+-	u32 remainder;
++	u32 remainder, remainder_minor;
+ 	char *name;
+ 	u64 tmp;
+ 
+@@ -145,6 +146,7 @@ static struct mtd_info *allocate_partiti
+ 		int i, max = parent->numeraseregions;
+ 		u64 end = child->part.offset + child->part.size;
+ 		struct mtd_erase_region_info *regions = parent->eraseregions;
++		uint32_t erasesize_minor = child->erasesize;
+ 
+ 		/* Find the first erase regions which is part of this
+ 		 * partition. */
+@@ -155,15 +157,24 @@ static struct mtd_info *allocate_partiti
+ 		if (i > 0)
+ 			i--;
+ 
+-		/* Pick biggest erasesize */
+ 		for (; i < max && regions[i].offset < end; i++) {
++			/* Pick biggest erasesize */
+ 			if (child->erasesize < regions[i].erasesize)
+ 				child->erasesize = regions[i].erasesize;
++			/* Pick smallest non-zero erasesize */
++			if ((erasesize_minor > regions[i].erasesize) && (regions[i].erasesize > 0))
++				erasesize_minor = regions[i].erasesize;
+ 		}
++
++		if (erasesize_minor < child->erasesize)
++			child->erasesize_minor = erasesize_minor;
++
+ 		BUG_ON(child->erasesize == 0);
+ 	} else {
+ 		/* Single erase size */
+ 		child->erasesize = master->erasesize;
++		if (master->erasesize_minor)
++			child->erasesize_minor = master->erasesize_minor;
+ 	}
+ 
+ 	/*
+@@ -171,26 +182,43 @@ static struct mtd_info *allocate_partiti
+ 	 * exposes several regions with different erasesize. Adjust
+ 	 * wr_alignment accordingly.
+ 	 */
+-	if (!(child->flags & MTD_NO_ERASE))
++	if (!(child->flags & MTD_NO_ERASE)) {
+ 		wr_alignment = child->erasesize;
++		if (IS_ENABLED(CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE) && child->erasesize_minor)
++			wr_alignment_minor = child->erasesize_minor;
++	}
+ 
+ 	tmp = mtd_get_master_ofs(child, 0);
+ 	remainder = do_div(tmp, wr_alignment);
+ 	if ((child->flags & MTD_WRITEABLE) && remainder) {
+-		/* Doesn't start on a boundary of major erase size */
+-		/* FIXME: Let it be writable if it is on a boundary of
+-		 * _minor_ erase size though */
+-		child->flags &= ~MTD_WRITEABLE;
+-		printk(KERN_WARNING"mtd: partition \"%s\" doesn't start on an erase/write block boundary -- force read-only\n",
+-			part->name);
++		if (wr_alignment_minor) {
++			tmp = mtd_get_master_ofs(child, 0);
++			remainder_minor = do_div(tmp, wr_alignment_minor);
++			if (remainder_minor == 0)
++				child->erasesize = child->erasesize_minor;
```

Comment:
Why do you set the `erasesize` here to `erasesize_minor`? That doesn't make sense to me...

Reply:
So that if a program uses the erasesize for a partition, to determine the minimum it needs to erase to rewrite a section of the parition, it does not get blocked by the (large) erasesize possibly exceeding the partition bounds.
Example: OpenWrt's `mtd`, in redboot FIS directory rewriting, uses the partition erasesize, but should probably use the partition size: https://github.com/openwrt/openwrt/pull/3271#issuecomment-696422410

Reply:
I see... Think this will take a little more time and maybe it makes sense to refactor the code into helper functions as it is not self-explaining and quite chaotic compared to other kernel code.

Reply:
I think this is fine as is, however a quick comment in the code might help make this clearer and save review time upstream.

Still, I wonder if you wouldn't want to adjust `wr_alignment` here: it doesn't make sense from a performance point of view to redo the whole division for the end of the partition if you already know that you are _starting_ on a minor boundary (although this is not in a perf-critical path).

Also correct me if I'm wrong but I don't think you need to fetch the master_ofs again: after the initial division (e.g. 64K EB), all you need to check is whether or not the remainder is a multiple of the minor erase since the minor erase is a multiple of the larger one. And you probably don't need the `remainder_minor` variable either.

All in all, assuming the above is correct, I'd write this like so (which saves a few checks and branches):

```c
tmp = mtd_get_master_ofs(child, 0);
remainder = do_div(tmp, wr_alignment);
if ((child->flags & MTD_WRITEABLE) && remainder) {
	if (wr_alignment_minor) {
		tmp = remainder;
		remainder = do_div(tmp, wr_alignment_minor);
		if (remainder == 0) {
			wr_alignment = wr_alignment_minor;
			/* Some programs check the partition erasize before writing, adjust it */
			child->erasesize = child->erasesize_minor;
		}
	}

	if (remainder != 0) {
		child->flags &= ~MTD_WRITEABLE;
		printk(KERN_WARNING"mtd: partition \"%s\" doesn't start on an erase/write block boundary -- force read-only\n",
			part->name);
	}
}
```

Likewise for the end (without the `wr_alignment` assignment of course).

Reply:
Note that a likely comment from upstream for the child->erasesize assignment may be that the kernel doesn't care about broken programs ;)

Reply:
> a quick comment in the code might help make this clearer

I have not worked out what I want it to say yet… Is this clear:?
Ensure a program which requests an (aligned) mtd_info->erasesize sized erase
will not be denied for the erase overflowing the partition end.

> adjust wr_alignment here: it doesn't make sense from a performance point of view to redo the whole division

Could break the `if (child->flags & MTD_WRITEABLE)` to above the division? One for start, and another for end.


> Also correct me if I'm wrong but I don't think you need to fetch the master_ofs again:

I might introduce another variable for this offset, rather than repeatedly calling it.

> after the initial division (e.g. 64K EB), all you need to check is whether or not the remainder is a multiple of the minor erase since the minor erase is a multiple of the larger one.

Yes, I expect that is the case, but I do know that is it guaranteed.

> And you probably don't need the remainder_minor variable either.

Yes, I like that, thanks.

Cheers

Reply:
> > a quick comment in the code might help make this clearer
> 
> I have not worked out what I want it to say yet… Is this clear:? Ensure a program which requests an (aligned) mtd_info->erasesize sized erase will not be denied for the erase overflowing the partition end.

How about: "Adjust partition erasesize: userspace requesting erasesize-sized erases for minor-aligned partitions will otherwise fail" ?

Still, I suspect this will trigger a NACK upstream, unless there's a very good reason the "program" cannot be fixed :)

> > adjust wr_alignment here: it doesn't make sense from a performance point of view to redo the whole division
> 
> Could break the `if (child->flags & MTD_WRITEABLE)` to above the division? One for start, and another for end.

I'm not sure I follow: if you succcessfully adjust the erasesize at "start", the next `& MTD_WRITEABLE` check is taken and the same computation is done for "end". If you didn't, then `& MTD_WRITEABLE` is not taken and the computation is not performed.

My suggestion was that in the first case, because you adjusted the erasesize at "start", it didn't make sense to first compute the "end" remainder on the larger erasesize which you already know you won't use (although it doesn't hurt). Or am I confused ?

> > Also correct me if I'm wrong but I don't think you need to fetch the master_ofs again:
> 
> I might introduce another variable for this offset, rather than repeatedly calling it.
> 
> > after the initial division (e.g. 64K EB), all you need to check is whether or not the remainder is a multiple of the minor erase since the minor erase is a multiple of the larger one.
> 
> Yes, I expect that is the case, but I do know that is it guaranteed.

Not sure what you mean? By (hardware) design eraseblocks are powers of two. A larger one is thus always a multiple of any smaller one.

> > And you probably don't need the remainder_minor variable either.
> 
> Yes, I like that, thanks.

np, yw :)

HTH